Fix to na.rm in decostand methods alr/clr

antagomir opened the pull request:
This fixes #661 

The `na.rm` option has been enabled in decostand methods "alr" and "clr".

Some notes:

1) "clr": With `na.rm=TRUE`, the NA entries in original data are replaced by 0 in the clr-transformed data; here zero is an arbitrary choice but no better solutions are available afaik, and the same approach is used in `compositions::clr` so at least we are in sync. With `na.rm=FALSE`, the entries that are NA in the original data will be also NAs in the clr-transformed data; the NAs are otherwise ignored in the calculations during the transformation. In the alternative implementation, `compositions::alr` NAs are always replaced with 0 (no na.rm option available). As far as I can see it is feasible to provide both options for `na.rm` for the "clr" method, we are providing that now.
 
2) "alr": similarly done for decostand "alr" transformation, user can choose whether to keep the NAs (`na.rm=FALSE`) or replace them by 0 (`na.rm=TRUE`). In comparison, I noticed that in the alternative implementation, `compositions::alr` NAs are always kept as NAs (no na.rm option available). As far as I can see it is feasible to provide both options for `na.rm` for the "alr" method, we are providing that now.

3) The "rclr" method will be handled separately in #619 

@TuomasBorman can you test and confirm that these are as you expected?

TuomasBorman commented on R/decostand.R:
```diff
@@ -156,61 +156,93 @@
     clog <- clog - means
+
+    # Replace missing values with 0
+    if (na.rm) {
```
You could add

`any(is.na(clog))`

as this is not affecting the result otherwise and the message could confuse

antagomir replied:
Good catch.

TuomasBorman commented on R/decostand.R:
```diff
@@ -225,6 +257,13 @@
     clog <- clog[, -reference] - refvector
+
+    # Replace missing values with 0
+    if (na.rm) {
```
Same here, consider adding `any(is.na(clog))`

TuomasBorman commented on R/decostand.R:
```diff
@@ -156,61 +156,93 @@
-        stop("'alr' cannot be used with negative data: use pseudocount >= ",
+    # Always na.rm=TRUE at this step
+    if (any(x <= 0, na.rm = TRUE)) {
+        stop("'alr' cannot be used with non-positive data: use pseudocount >= ",
```
"use pseudocount >=". This is wrong. It should be "use pseudocount >"

TuomasBorman reviewed (approved):
I think now works as expected. See the comments.

antagomir commented on R/decostand.R:
```diff
@@ -156,61 +156,93 @@
 
 ## Modified from the original version in mia R package
 .calc_clr <-
-    function(x, pseudocount=0, na.rm = TRUE)
+    function(x, na.rm, pseudocount=0)
 {
```
done

antagomir commented on R/decostand.R:
```diff
@@ -156,61 +156,93 @@
 
+
+
 .calc_alr <-
```
ok

antagomir commented:
Great, fixed the remarks.

antagomir commented:
This PR implements updated versions for:
- Robust CLR `decostand(..., method="rclr")`
- Robust Aitchison distance `vegdist(..., method="robust.aitchison")`

This PR is now updated with the following:
- Earlier separate PR #619 has been merged here, taking into account the feedback there (I hope the formatting issues are now ok, those seemed to be the main remarks)
- Approach is now synced with the Python implementation
- Handling of missing values has been improved.


One major point that I wanted to check before giving the very final touch on this PR: can we export the new function ROptSpace? This is not only used to calculate the robust.aitchison distance in vegdist but it is also more widely applicable and this would make the methods directly comparable with their equivalents in Python/gemelli, supporting cross-testing etc.

The OptSpace function is used for matrix completion after the rclr transformation. The rclr transformation usually results in some NA values in the transformed matrix, and the matrix completion is commonly done after this step. The reason for separating the rclr transformation and the matrix completion for missing values is that the matrix completion is not used on the rclr-transformed matrix directly; instead it returns other (SVD-related) outputs that are needed in further downstream analyses (e.g. robust Aitchison distance, robust PCA). The OptSpace matrix completion is used in the robust.aitchison distance in vegdist. 

The new OptSpace function is a modified version of ROptSpace::OptSpace function, with modifications to better fit ecological data. The original implementation has MIT license and it has been cited properly. This is compatible with the GPL2 license of vegan and can be included this way. 

So I would like to confirm if it is OK to keep the OptSpace function here I can make the final checks, otherwise I would remove it before declaring the PR complete.

antagomir commented:
Some minor updates added. 

This PR accommodates all previous feedback. 

It is ready and preliminary tests against the Python/Gemelli implementation have been done and the results are aligned.

If possible, we hope to keep `OptSpace` as an exported function as it has further use (e.g. robust PCA).

cameronmartino commented:
Hi,

I am the author of the method being added here & the python implementation (diecode now moved to [gemelli](https://github.com/biocore/gemelli)). Greatly appreciate @antagomir adding this implementation in R! I have been in some conversation with him on this offline. To help push the PR along, I checked the implementation in this PR and the results are now the same as the python implementation (as pointed out in issue #570 they did not match before). Simple test example below. Hope this helps get this PR through, there has been the demand in R but I didn't have the bandwidth to tackle this myself.

First, grab some (small) real data as an example
```bash
wget https://github.com/biocore/gemelli/raw/refs/heads/master/ipynb/tutorials/qiime2-moving-pictures-tutorial/table.biom
# export to tsv for R
biom convert -i table.biom -o table.tsv --to-tsv
sed -i '' 1d table.tsv 
````

Generate the distances and ordination in R (using this PR branch)

```R
library(devtools)
install_github("antagomir/vegan", ref = "nafix")
library(vegan)
# load table
table <- read.table("table.tsv", header = T, comment = '', check = F, sep = '\t', row.names = 1)
table <- t(as.matrix(table))
rclr_table <- vegan::decostand(table, method="rclr")
# calculate the distance
distance = vegan::vegdist(rclr_table, method = "euclidean")
write.table(as.matrix(distance), file = "R-rpca-distances.tsv", row.names=rownames(table), col.names=rownames(table), sep="\t")
# save to test RPCA
svd.res <- svd(rclr_table)
write.table(svd.res$u, file = "R-rpca-sample-loadings.tsv", row.names=rownames(table), sep="\t")
write.table(svd.res$d, file = "R-rpca-eig-vals.tsv", row.names=FALSE, sep="\t")
write.table(svd.res$v, file = "R-rpca-feature-loadings.tsv", row.names=colnames(table), sep="\t")
```

Generate the distances and ordination in python.

```python
import numpy as np
import pandas as pd
from biom import load_table
from gemelli.rpca import rpca
from gemelli import __version__
print('Version: %s' % __version__)

# load table
bt = load_table('table.biom')
# run RPCA (this runs RCLR internally)
py_ordination, py_distance = rpca(bt)
# extract RPCA results from skbio type
py_rpca_U = py_ordination.samples.values
py_rpca_V = py_ordination.features.values
py_rpca_S = py_ordination.eigvals.values
py_dist = py_distance.to_data_frame().values
# check the OptSpace results
R_rpca_U = pd.read_csv('R-rpca-sample-loadings.tsv', sep='\t', index_col=None).values[:, :3]
R_rpca_S = pd.read_csv('R-rpca-eig-vals.tsv', sep='\t', index_col=None).values[:3]
R_rpca_V = pd.read_csv('R-rpca-feature-loadings.tsv', sep='\t', index_col=None).values[:, :3]
R_rpca_dist = pd.read_csv('R-rpca-distances.tsv', sep='\t', index_col=None).values
# check
print(' Sample loadings match: %s \n' % str(abs((np.abs(R_rpca_U) - np.abs(py_rpca_U)).sum()) < 1e-6),
        'Feature loadings match: %s \n' % str(abs((np.abs(R_rpca_V) - np.abs(py_rpca_V)).sum()) < 1e-6),
        'Singular values match: %s \n' % str(abs((np.abs(R_rpca_S) - np.abs(py_rpca_S)).sum()) < 1e-6),
        'Distances match: %s \n' % str(abs((R_rpca_dist - R_rpca_dist).sum()) < 1e-6))
```
```output
Version: 0.0.12
 Sample loadings match: True 
 Feature loadings match: True 
 Singular values match: True 
 Distances match: True 
```

antagomir commented:
This is now thoroughly tested and cross-validated with an independent implementation and as far as I can see, we have fixed all related issues. Would the start of 2025 be a good time to merge the PR? ping @jarioksa

antagomir commented:
Dear @jarioksa @gavinsimpson - there has been general interest towards the robust CLR and robust Aitchison. These methods are already in vegan, following our previous PR. 

This present PR includes fixes to the previously reported issue + improved implementation, and it has been extensively verified with an independent Python implementation by @cameronmartino

If there are any issues remaining towards accepting this improved update in vegan we will be happy to work through them. As far as I can see, all previous feedback has been incorporated.

gavinsimpson commented on R/decostand.R:
```diff
@@ -269,3 +311,323 @@
+  eps <- nnZ.E/sqrt(m*n)
+  
+  ## Preprocessing : ROPT  : implied rank
+  if (is.na(ROPT)){
```
I don't like this design; Setting an argument to be `NA` implies that its value is *missing*. Better conventions are `NULL`. However, in this case, I think we should define `ropt = TRUE` as the argument and default value. Then in the function we would do

```r
if (ropt) {
  rpot <- 3 # set your default
}
```

and then we can document `ropt` to take a logical `TRUE/FALSE` or a positive integer for the rank.

gavinsimpson commented on R/decostand.R:
```diff
@@ -269,3 +311,323 @@
+  ## Preprocessing : ROPT  : implied rank
+  if (is.na(ROPT)){
+    if (verbose){
+      cat("* OptSpace: Guessing an implicit rank.")
```
This should be `message()`

gavinsimpson commented on R/decostand.R:
```diff
@@ -269,3 +311,323 @@
+    }
+    r <- min(max(round(.guess_rank(M_E, nnZ.E)), 2), m-1)
+    if (verbose){
+      cat(paste0('* OptSpace: Guessing an implicit rank: Estimated rank : ',r))
```
`message()` not `cat()`

gavinsimpson commented on R/decostand.R:
```diff
@@ -269,3 +311,323 @@
+    }
+  } else {
+    r <- round(ROPT)
+    if ((!is.numeric(r)) || (r<1) || (r>m) || (r>n)){
```
spacing around operators and before `{`

gavinsimpson commented on R/decostand.R:
```diff
@@ -269,3 +311,323 @@
+  }
+  
+  ## Preprocessing : NITER : maximum number of iterations
+  if ((is.infinite(NITER))||(NITER<=1)||(!is.numeric(NITER))){
```
spacing around everything

gavinsimpson commented on R/decostand.R:
```diff
@@ -269,3 +311,323 @@
+  rho <-  eps*n
+  
+  ## Main Computation
+  rescal_param <- sqrt(nnZ.E*r/(norm(M_E,'f')^2))
```
Spacing

gavinsimpson commented on R/decostand.R:
```diff
@@ -269,3 +311,323 @@
+    Z <- tmpgrad$Z
+    # line search for the optimum jump length
+    t <- .aux_getoptT(X, W, Y, Z, S, M_E, E, m0, rho)
+    X <- X + t*W;
```
no trailing `;`

gavinsimpson commented on R/decostand.R:
```diff
@@ -269,3 +311,323 @@
+    # line search for the optimum jump length
+    t <- .aux_getoptT(X, W, Y, Z, S, M_E, E, m0, rho)
+    X <- X + t*W;
+    Y <- Y + t*Z;
```
no trailing `;`

gavinsimpson commented on R/decostand.R:
```diff
@@ -269,3 +311,323 @@
+  out$Y <- Y
+  out$dist <- dist
+  if (verbose){
+    cat('* OptSpace: estimation finished.')
```
`message()`

gavinsimpson commented on R/decostand.R:
```diff
@@ -269,3 +311,323 @@
+  }
+  NITER <- round(NITER)
+  
+  m0 <- 10000
```
what is this constant for?

antagomir replied:
Scaling parameter for gradient calculation. I added a small comment for this internal parameter. References for the method have been provided on the manpage.

gavinsimpson commented on R/decostand.R:
```diff
@@ -269,3 +311,323 @@
+  idxfind <- (z < 1)
+  y[idxfind] <- 0
+  out <- sum(y)
+  return(out)
```
It is better style now to not use `return` as R automagically calls `return()` on the value of the last line of a function. Just leave this as `out`.

gavinsimpson commented on R/decostand.R:
```diff
@@ -269,3 +311,323 @@
+  out2 <- rho*.aux_G(Y,m0,r)
+  out3 <- rho*.aux_G(X,m0,r)
+  out  <- out1+out2+out3
+  return(out)
```
It is better style now to not use `return` as R automagically calls `return()` on the value of the last line of a function. Just leave this as `out`.

gavinsimpson commented on R/decostand.R:
```diff
@@ -269,3 +311,323 @@
+
+  # -------------------------------------------
+
+  return(out)
```
It is better style now to not use `return` as R automagically calls `return()` on the value of the last line of a function. Just leave this as `out`.

gavinsimpson replied:
It is better style now to not use `return` as R automagically calls `return()` on the value of the last line of a function. Just leave this as `out`.

gavinsimpson commented on R/decostand.R:
```diff
@@ -156,61 +156,96 @@
+     xx <- M_I
+   }
+  
+   return(xx)
```
It is better style now to not use `return` as R automagically calls `return()` on the value of the last line of a function. Just leave this as `out`.

gavinsimpson commented on R/decostand.R:
```diff
@@ -269,3 +311,323 @@
+  resgrad <- list()
+  resgrad$W <- W
+  resgrad$Z <- Z
+  return(resgrad)
```
Just `resgrad`.

gavinsimpson commented on R/decostand.R:
```diff
@@ -269,3 +311,323 @@
+  
+  S <- solve(A, C)
+  out <- matrix(S, nrow=r)
+  return(out)
```
`out`

gavinsimpson commented on R/decostand.R:
```diff
@@ -269,3 +311,323 @@
+
+# Aux 5 : optimal line search ---------------------------------------------
+# @keywords internal
+.aux_getoptT <- function(X, W, Y, Z, S, M_E, E, m0, rho)
```
These arguments need comments to help us understand what they are for - will help with ongoing maintenance of the code

gavinsimpson commented on R/decostand.R:
```diff
@@ -269,3 +311,323 @@
+
+# Aux 4 : Sopt given X and Y ----------------------------------------------
+# @keywords internal
+.aux_getoptS <- function(X, Y, M_E, E)
```
These arguments need comments to help us understand what they are for - will help with ongoing maintenance of the code

antagomir replied:
i added a short summary on all

gavinsimpson commented on R/decostand.R:
```diff
@@ -269,3 +311,323 @@
+  out <- (X * matrix(z, nrow=nrow(X), ncol=ncol(X), byrow=FALSE))/(m0*r)
+}
+# @keywords internal
+.aux_gradF_t <- function(X, Y, S, M_E, E, m0, rho)
```
These arguments need comments to help us understand what they are for - will help with ongoing maintenance of the code

antagomir replied:
i added a short summary on all

gavinsimpson commented on R/decostand.R:
```diff
@@ -269,3 +311,323 @@
+
+# Aux 3 : compute the gradient --------------------------------------------
+# @keywords internal
+.aux_Gp <- function(X,m0,r)
```
These arguments need comments to help us understand what they are for - will help with ongoing maintenance of the code.

And spaces after `,`

antagomir replied:
i added a short summary on all

gavinsimpson commented on R/decostand.R:
```diff
@@ -269,3 +311,323 @@
+    t <- t/2
+  }
+  out <- t
+  return(t)
```
`t`

gavinsimpson commented on R/vegdist.R:
```diff
@@ -48,7 +48,7 @@
         x <- decostand(x, "clr", ...)  # dots to pass possible pseudocount
     if (method == 22)  # robust.aitchison
-        x <- decostand(x, "rclr") # No pseudocount for rclr
+        x <- decostand(x, "rclr", nr.rm=na.rm, ...) # No pseudocount for rclr
```
Is this a typo? Should it be `na.rm = na.rm, ...`?

antagomir replied:
fixed

gavinsimpson commented on man/OptSpace.Rd:
```diff
@@ -0,0 +1,77 @@
+\title{OptSpace: algorithm for matrix reconstruction from a partially revealed set}
+\description{
+This function was adapted from the original source code in the
+ROptSpace R package (version 0.2.3; MIT License) by Raghunandan
```
Use `\pkg{package_name}` to markup an R package name

gavinsimpson commented on man/OptSpace.Rd:
```diff
@@ -0,0 +1,77 @@
+This function was adapted from the original source code in the
+ROptSpace R package (version 0.2.3; MIT License) by Raghunandan
+H. Keshavan, Andrea Montanari, Sewoong Oh (2010). See
+ROptSpace::OptSpace for more information.  Let's assume an ideal
```
`ROptSpace::OptSpace` -> `\code{ROptSpace::OptSpace}`, assuming we don't want to link to the function in the package (if we do then the package would have to be `Suggests`, which I think we wanted to avoid?)

gavinsimpson commented on man/OptSpace.Rd:
```diff
@@ -0,0 +1,77 @@
+C(\alpha)\sqrt{nr/|E|}}.
+}
+\usage{
+OptSpace(x, ROPT=3, NITER=5, TOL=1e-5, verbose=FALSE)
```
update these args when the function switches to lower case argument names

gavinsimpson commented on man/OptSpace.Rd:
```diff
@@ -0,0 +1,77 @@
+\examples{
+data(varespec)
+# rclr transformation with no matrix completion for the 0/NA entries
+x <- decostand(varespec, method="rclr", impute=FALSE)
```
spaces around `=`

gavinsimpson commented on man/OptSpace.Rd:
```diff
@@ -0,0 +1,77 @@
+# rclr transformation with no matrix completion for the 0/NA entries
+x <- decostand(varespec, method="rclr", impute=FALSE)
+# Add matrix completion
+xc <- OptSpace(x, ROPT=3, NITER=5, TOL=1e-5, verbose=FALSE)$M
```
spaces around `=`

gavinsimpson commented on man/decostand.Rd:
```diff
@@ -106,25 +106,37 @@ decobackstand(x, zap = TRUE)
      adding it manually to the input data, or by using the argument
      \code{pseudocount} as in
-     \code{decostand(x, method = "clr", pseudocount = 1)}. Adding
+     \code{decostand(x, method = "clr", na.rm=TRUE, pseudocount = 1)}. Adding
```
`na.rm = TRUE,`

gavinsimpson commented on man/decostand.Rd:
```diff
@@ -106,25 +106,37 @@ decobackstand(x, zap = TRUE)
+     sample-wide values \eqn{x} that are positive (> 0). The OptSpace
+     algorithm performs matrix completion for the missing values
+     that result from log transformation of the zero entries in the
+     original input data. The vegan implementation of the matrix
```
I don't think we need to repeat this
>The vegan implementation of the matrix
     completion was adapted from the original implementation
     ROptSpace::OptSpace (version 0.2.3) following Keshavan et
     al. (2010).

here. Just direct people to the OptSpace Rd file with `See \link[vegan]{OptSpace} for more details.` Or something like that.

gavinsimpson commented on man/decostand.Rd:
```diff
@@ -204,13 +220,17 @@ sptrans <- wisconsin(varespec)
 varespec.clr <- decostand(varespec, "clr", pseudocount=1)
 
+# Robust CLR (rclr) transformation for rows, no pseudocount necessary
+varespec.rclr <- decostand(varespec, "rclr", impute=TRUE)
```
`impute = TRUE`

gavinsimpson commented on man/vegdist.Rd:
```diff
@@ -419,5 +419,8 @@ vare.dist <- vegdist(decostand(varespec, "norm"), "euclidean")
 # Range standardization with "altGower" (that excludes double-zeros)
 vare.dist <- vegdist(decostand(varespec, "range"), "altGower")
+# Robust Aitchison distance equals to Euclidean distance for rclr transformed data
+vare.dist <- vegdist(decostand(varespec, "rclr"), method="euclidean")
```
`method = "euclidean"`

gavinsimpson commented on tests/decostand-tests.R:
Ideally, this file would have similar standards to the rest of vegan, with spacing around operators, like `=`, etc. But I don't think it needs doing personally. @jarioksa may differ.

antagomir replied:
done

gavinsimpson commented:
@antagomir I have done a code review - mostly trivial for style, but matching vegan-style is needed to help us maintain code. I found one typo-bug. Some comments and suggestions to change the code and documentation in places.

antagomir commented on R/decostand.R:
```diff
@@ -156,61 +156,96 @@
+
+    # Replace missing values with 0
+    if (na.rm && any(is.na(clog))) {
+        cat("Replacing missing values with zero for clr. You can disable this with na.rm=FALSE.")    
```
Right - I was somehow mistaken that "cat" would be preferred in vegan. I agree, we go with `message()`.

Unless I am mistaken, the decostand default is na.rm=FALSE (in [decostand.R, line 2](https://github.com/vegandevs/vegan/blob/186573d038a6f7ece120a3d5af5e5eb2545146a6/R/decostand.R#L2))?

Agreed with warning and formatting & updated now.

gavinsimpson replied:
> Unless I am mistaken, the decostand default is na.rm=FALSE

Right, sorry. Confused myself just scrolling through the diff.

antagomir replied:
Ok good. Warning is warranted in any case. Added.

antagomir commented:
Ok I have resolved all comments. I tried to check all files comprehensively for similar issues, I hope I covered all. 

The tests seem to pass ok.

Cameron Martino has substantially contributed to the implementation. I am not sure what the policy is but you can consider adding him in the DESCRIPTION file (or I can add it in PR if agreed so). 

Ready to merge? If anything else remains just let me know and we will do asap.

gavinsimpson commented:
@antagomir I don't see that my comments have been resolved - there are still trailing `;` in places, and using `return(foo)` as the last statement in a function is redundant, just use `foo`. Those were just a few things that I saw in `decostand.R` that are outstanding; I didn't check all the comments, those were ones that I noticed and decided to wait until they are complete before signing off.

antagomir commented on R/decostand.R:
```diff
@@ -269,3 +311,323 @@
+    stop("* OptSpace : no infinite value in x is allowed")
+  }
+  if (!any(is.na(x))){
+    stop("* OptSpace : there is no unobserved values as NA")
```
true, good catch.

antagomir commented:
Thanks!

I had resolved the comments as I went through and solved them. I have now reopened them for checking.

Github had hidden your 20+ comments that needed to be separately clicked open ("show more"). I had somehow not noticed that and mistakingly thought that I already went through all issues. Now I saw those, including trailing semicolon and foo etc. I have addressed all feedback now but there is still something that escaped my attention just let me know.

This PR is ready from my side but ready to add fixes as necessary.

antagomir commented:
The GHA fails but the issue seems unrelated to this PR - I also get the same "scores" error when I take a fresh clone from vegandevs/vegan master.

jarioksa commented:
@antagomir the error on missing `scores` **is** related to this PR: you add function `optspace`, but in `NAMESPACE` you export function `OptSpace`. Exporting a non-existent function ruins parsing the `NAMESPACE` exports and this first emerges when you try to build dynamic issues in help page `man/plot.cca.Rd`. Fixing the `NAMESPACE` will solve this issue.

antagomir commented:
Ahh, yes. The NAMESPACE of course.

Should be ok now?

gavinsimpson reviewed (approved):
There's some trivial things that I might clean up, but I can do that during the merge. Thanks @antagomir

gavinsimpson commented on tests/decostand-tests.R:
```diff
@@ -1,50 +1,89 @@
+
+# Test that NAs are handled as expected in CLR
+x <- testdata
+x[sample(prod(dim(x)), 50)] <- NA # insert some NAs in the data
```
Should this have a seed set before we call `sample()`?

gavinsimpson replied:
As you use `testdata` much later in L71 and L72 and dump the output to the test log we need to make this a fixed seed I think otherwise every run gives different results.

gavinsimpson commented on tests/decostand-tests.R:
```diff
@@ -1,50 +1,89 @@
+max(abs(na.omit(decostand(x, "alr", na.rm = FALSE, pseudocount = 1)[inds] -
+                decostand(x, "alr", na.rm = TRUE, pseudocount = 1)[inds]))) < 1e-6
+
+# Test that NAs are handled as expected in ALR
```
I'm seeing problems with this section

gavinsimpson commented on tests/decostand-tests.R:
```diff
@@ -1,50 +1,89 @@
+all(is.na(decostand(x, "clr", na.rm = FALSE, pseudocount = 1)[is.na(x)])) == TRUE # NAs
+# For the other (non-NA) values, we get non-NA values back
+any(is.na(decostand(x, "clr", na.rm = FALSE, pseudocount = 1)[!is.na(x)])) == FALSE
+any(is.na(decostand(x, "clr", MARGIN = 2, na.rm = FALSE, pseudocount = 1)[!is.na(t(x))])) == FALSE 
```
This is not returning `TRUE`:

```{r}
> any(is.na(decostand(x, "clr", MARGIN = 2, na.rm = FALSE, pseudocount = 1)[!is.na(t(x))])) == FALSE
[1] FALSE
```

Is this expected?

antagomir replied:
Despite all the efforts to check each line, this had somehow escaped my attention.. good catch. This should be

`any(is.na(decostand(x, "clr", MARGIN = 2, na.rm = FALSE, pseudocount = 1)[!is.na(x)])) == FALSE` 

i.e. no transpose around `is.na(x)`.

antagomir replied:
You can change, or I update the PR?

gavinsimpson replied:
I'll change; I have a few more lines to check with you that are causing some output diffs for me but I won't get to those until later this evening or first thing tomorrow.

gavinsimpson commented on tests/decostand-tests.R:
```diff
@@ -53,35 +92,52 @@ test2 <- test2[, -1]
 # Expect high correlation
 cor(unlist(test), unlist(test2)) > 0.99
```
The comment above suggests that we should see some `TRUE` values here but in my testing, all of these correlations are *not* $>0.99$. So I'm not sure what this section is supposed to be testing. It would be good to have an actual expectation to compare with, but clarifying whether something needs to be changed here would be a start?

antagomir replied:
This test was related to the earlier non-imputation version and can be removed.

gavinsimpson replied:
Great - I'll make that change locally.

gavinsimpson commented on tests/decostand-tests.R:
```diff
@@ -53,35 +92,52 @@ test2 <- test2[, -1]
+all(as.matrix(x1) == as.matrix(x2c))
+
+x2c <- optspace(x2, ropt = 1, niter = 5, tol = 1e-5, verbose = FALSE)$M
+all(as.matrix(x1) == as.matrix(x2c))
```
This is returning `FALSE` for me, but testing for exact equality is not very useful here. With `all.equal()`, I'm seeing

```r
> all.equal(as.matrix(x1), as.matrix(x2c))
[1] "Attributes: < Names: 1 string mismatch >"                                         
[2] "Attributes: < Component 2: Modes: list, numeric >"                                
[3] "Attributes: < Component 2: Length mismatch: comparison on first 2 components >"   
[4] "Attributes: < Component 2: Component 1: target is NULL, current is numeric >"     
[5] "Attributes: < Component 2: Component 2: Modes: character, numeric >"              
[6] "Attributes: < Component 2: Component 2: Lengths: 44, 1 >"                         
[7] "Attributes: < Component 2: Component 2: target is character, current is numeric >"
[8] "Mean relative difference: 0.9140407"
```

where the last line shows a significant deviation between `x1` and `x2c` - is this to be expected or not?

antagomir replied:
This one (with ropt = 1); lines 106-107 can be removed. These rows might have been left there by accident as x1 and x2c are allowed differ here.

antagomir commented:
Ok I have responded now. My bad overlooking those two cases, I hope all ok now.

gavinsimpson commented:
Thanks @antagomir all merged now

jarioksa commented:
This was much bigger than I anticipated. Thanks for the hard work @antagomir and @gavinsimpson. Now only two things are missing: NEWS entry and updating authorship in DESCRIPTION.

It took a while to get this merged, but the main thing is how fast this goes to CRAN. I'm going to start working on a major CRAN release (2.7-x). There are a few things that must be solved, and I'm going to have a new issue on release plan.
